node/pod.go

```diff
@@ -156,40 +156,25 @@ func (pc *PodController) handleProviderError(ctx context.Context, span trace.Spa
 	span.SetStatus(origErr)
 }
 
-func (pc *PodController) deletePod(ctx context.Context, namespace, name string) error {
+func (pc *PodController) deletePod(ctx context.Context, pod *corev1.Pod) error {
 	ctx, span := trace.StartSpan(ctx, "deletePod")
 	defer span.End()
-
-	pod, err := pc.provider.GetPod(ctx, namespace, name)
-	if err != nil {
-		if errdefs.IsNotFound(err) {
-			// The provider is not aware of the pod, but we must still delete the Kubernetes API resource.
-			return pc.forceDeletePodResource(ctx, namespace, name)
-		}
-		return err
-	}
-
-	// NOTE: Some providers return a non-nil error in their GetPod implementation when the pod is not found while some other don't.
-	if pod == nil {
-		// The provider is not aware of the pod, but we must still delete the Kubernetes API resource.
-		return pc.forceDeletePodResource(ctx, namespace, name)
-	}
-
 	ctx = addPodAttributes(ctx, span, pod)
 
-	var delErr error
-	if delErr = pc.provider.DeletePod(ctx, pod.DeepCopy()); delErr != nil && !errdefs.IsNotFound(delErr) {
-		span.SetStatus(delErr)
-		return delErr
+	err := pc.provider.DeletePod(ctx, pod.DeepCopy())
+	if errdefs.IsNotFound(err) {
+		// TODO: Figure out what to do here. Previously, we used to delete the pod non-gracefully from Kubernetes
+		// The problem is that if a pod was deleted, and then it is no longer in the provider, we end up deleting it
+		// from kubernetes due to no reason whatsoever.
```

Comment:
This is confusing to me. _if the pod is deleted_ means the pod was marked for deletion in the API, right? After all, `deletePod` is called when an event for deletion is received. So, if it's marked for deletion and it's not found in the provider, why shouldn't we (finally) delete it from the API?

Reply:
@pires IMHO, that's erroneous behaviour. Imagine this case:

1. The user sets deletion grace period to 60 seconds
2. The VK receives this, and calls delete on provider
3. The provider deletes the pod, and NotifyPod with a terminal status (but the update is delayed)
4. The user cranks deletion timestamp down to 30 seconds
5. The VK receives this update, and calls deletepod on the provider, and gets a pod not found
6. ???

If ??? is delete the pod from API server, we will never get the update in step 3.

Reply:
I kind of wish that Kubernetes had vector clocks. So, you could say, "Does this object have this update?" If it's strictly newer, then do the delete.

Reply:
Another wrench here, whether or not `DeletePod` should even return a `NotFound` is not straight forward.

Reply:
I think we may have different understandings of grace-period here. The VK, at most, should wait no more than grace-period duration for the provider to announce the pod was in fact deleted, hence the responsibility of deleting it, even without knowing the grace-period, is provider's only.

Maybe [this](https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-terminating-with-grace) helps us being on the same page?

Reply:
If the contention is this comment, then I can remove it.

--

`terminationGracePeriodSeconds`
> Optional duration in seconds the pod needs to terminate gracefully. May be decreased in delete request. Value must be non-negative integer. The value zero indicates delete immediately. If this value is nil, the default grace period will be used instead. The grace period is the duration in seconds after the processes running in the pod are sent a termination signal and the time when the processes are forcibly halted with a kill signal. Set this value longer than the expected cleanup time for your process. Defaults to 30 seconds.

If the provider gets a DeletePod, nothing in our specification states that the pod will actually be deleted by the time the call returns. 

As stated in the API specification:
> May be decreased in delete request.

If the terminationgraceperiodseconds is cranked down, and the provider has already deleted the pod, this can result in us erroneously deleting the pod in K8s prior to the expiration of the grace period.

Reply:
OK maybe we're missing each other. `kubectl delete --grace-period` overrides `PodSpec.terminationGracePeriodSeconds`. The VK should enforce the former while the provider should enforce the latter, or so is my understanding.

Reply:
Right. But if termination grace period changes, we need to inform the provider (which we do), but it can result in an erroneous action being taken. One of the things we can do is say if it is not found AND it was because this was a duplicate delete, ignore not found.

Reply:
I don't think we can be sure about that. Say the VK crashes/restarts in between events. It will lose its in-memory map (that you're adding in this PR) and won't understand it's a duplicate delete, right?

Reply:
@pires You're right. I believe that behaviour is erroneous. I believe we need a mechanism around that in the future.